Detectors/TOF/reconstruction/src/Clusterer.cxx

```diff
@@ -0,0 +1,199 @@
+  //setL0L1Latency(); // to be filled (maybe)
+  //setDeltaBC(); // to be filled (maybe)
+
+  int chan1, chan2;
+  int phi1, phi2;
+  int eta1, eta2;
+  int deltaPhi, deltaEta;
+  int mask;
+
+  mContributingDigit[0]->getPhiAndEtaIndex(phi1, eta1);
+  // now set the mask with the secondary digits
+  for (int idig = 1; idig < mNumberOfContributingDigits; idig++) {
+    mContributingDigit[idig]->getPhiAndEtaIndex(phi2, eta2);
+    deltaPhi = phi1 - phi2;
+    deltaEta = eta1 - eta2;
+    if (deltaPhi == 1) {   // the digit is to the LEFT of the cluster; let's check about UP/DOWN/Same Line
+      if (deltaEta == 1) { // the digit is DOWN LEFT wrt the cluster
+        mask = Cluster::kDownLeft;
+      } else if (deltaEta == -1) { // the digit is UP LEFT wrt the cluster
+        mask = Cluster::kUpLeft;
+      } else { // the digit is LEFT wrt the cluster
+        mask = Cluster::kLeft;
+      }
+    } else if (deltaPhi == -1) { // the digit is to the RIGHT of the cluster; let's check about UP/DOWN/Same Line
+      if (deltaEta == 1) {       // the digit is DOWN RIGHT wrt the cluster
+        mask = Cluster::kDownRight;
+      } else if (deltaEta == -1) { // the digit is UP RIGHT wrt the cluster
+        mask = Cluster::kUpRight;
+      } else { // the digit is RIGHT wrt the cluster
+        mask = Cluster::kRight;
+      }
+    } else if (deltaPhi == 0) { // the digit is on the same column as the cluster; is it UP or Down?
+      if (deltaEta == 1) {      // the digit is DOWN wrt the cluster
+        mask = Cluster::kDown;
+      } else if (deltaEta == -1) { // the digit is UP wrt the cluster
+        mask = Cluster::kUp;
+      } else { // impossible!!
+        LOG(DEBUG) << " Check what is going on, the digit you are trying to merge to the cluster must be in a different channels... " << FairLogger::endl;
+      }
+    } else { // impossible!!! We checked above...
+      LOG(DEBUG) << " Check what is going on, the digit you are trying to merge to the cluster is too far from the cluster, you should have not got here... " << FairLogger::endl;
+    }
+    c.addBitInContributingChannels(mask);
+  }
+
+  // filling the MC labels of this cluster; the first will be those of the main digit; then the others
+  if (digitMCTruth != 0x0) {
+    int lbl = mClsLabels->getIndexedSize(); // this should correspond to the number of digits also;
+    for (int i = 0; i < mNumberOfContributingDigits; i++) {
+      int digitLabel = mContributingDigit[i]->getLabel();
+      gsl::span<const o2::MCCompLabel> mcArray = digitMCTruth->getLabels(digitLabel);
+      for (int j = 0; j < static_cast<int>(mcArray.size()); ++j) {
+        auto evID = digitMCTruth->getElement(digitMCTruth->getMCTruthHeader(digitLabel).index + j).getEventID();
+        auto trID = digitMCTruth->getElement(digitMCTruth->getMCTruthHeader(digitLabel).index + j).getTrackID();
+        auto srcID = digitMCTruth->getElement(digitMCTruth->getMCTruthHeader(digitLabel).index + j).getSourceID();
```

Comment:
this looks a bit too verbose. Consider using the much shorter variant `auto label = digitMCTruth->getElement()` and saying `mClsLabels->addElement(lbl, label);`